### PR TITLE
Decide if Zebra is at the chain tip

### LINKED ISSUE #2592: Activate Mempool storage based on a list of recent syncer obtain/extend lengths
## Motivation

We want to control when we activate mempool, for testing purposes as well as when zebra reaches the chain tip. 

## Testing

Test that activation happens near the chain tip

## Related Work

This change depends on ticket #2595.

### PR DESCRIPTION
**Pull request by upbqdn**
## Motivation

Zebra needs a way to tell when it has reached the chain tip.

Closes #2592.

## Solution

This PR simply calculates the average recent sync length, and returns `true` if this value falls below 20. This constant (20) is based on https://github.com/ZcashFoundation/zebra/issues/2592#issuecomment-897304684. 

### Reviewer Checklist

  - [ ] Code implements the behavior required in #2592

### CODE REVIEW

**Comment by upbqdn** [comment]
The function `is_close_to_tip` currently returns `true` even when there are too many zero values returned by `RecentSyncLengths`.  I left it this way because I'm not sure how `RecentSyncLengths` really behaves when Zebra is synchronized.

Also note that `is_close_to_tip` can return `true` only if there is no value greater than 80 in the list of recent sync lengths since Average([80, 0, 0, 0]) = 20.

**Review by teor2345 — COMMENTED** [review]
This is a good minimal implementation.
We can tweak the constants later during testing, if needed.

It looks like the `waits_until_close_to_tip` test is failing because it assumes that a random series of sync lengths will always be close to the tip. (Which isn't true.)

And in this PR we actually check the lengths, so the test fails.

@jvff should be able to help out here.

**Comment by oxarbitrage** [comment]
I think this looks pretty good. Should we create a specific test for it or our current tests covers it ?

**Comment by teor2345** [comment]
> I think this looks pretty good. Should we create a specific test for it or our current tests covers it ?

I think the proptest is good, but I'd also like us to test some test vectors.

For example:
1. totally empty sync lengths array is not near tip
2. sync lengths array with all zeroes is near tip
3. sync lengths array with all `500`s is not near tip

Those are 3 basic tests that should keep on working, even if we tweak the algorithm or constants.

**Review by jvff — COMMENTED** [review]
Looks good! If you need any help adding the test vectors, let me know :+1:

File: zebrad/src/components/sync/status/tests.rs [hunk 1]
```diff
@@ -138,3 +138,35 @@ proptest! {
         }
     }
 }
+
+/// Test if totally empty sync lengths array is not near tip.
+#[test]
+fn empty_sync_lengths() {
+    let (status, _recent_sync_lengths) = SyncStatus::new();
+
+    assert!(status.is_close_to_tip() == false);
```
Comment (oxarbitrage) [inline zebrad/src/components/sync/status/tests.rs]:
```suggestion
    assert!(!status.is_close_to_tip());
```
clippy

File: zebrad/src/components/sync/status/tests.rs [hunk 2]
```diff
@@ -138,3 +138,35 @@ proptest! {
...
+/// Test if sync lengths array with all zeroes is near tip.
+#[test]
+fn zero_sync_lengths() {
+    let (status, mut recent_sync_lengths) = SyncStatus::new();
+
+    for _ in 0..RecentSyncLengths::MAX_RECENT_LENGTHS {
+        recent_sync_lengths.push_extend_tips_length(0);
+    }
+
+    assert!(status.is_close_to_tip() == true);
```
Comment (oxarbitrage) [inline zebrad/src/components/sync/status/tests.rs]:
```suggestion
    assert!(status.is_close_to_tip());
```
clippy

File: zebrad/src/components/sync/status/tests.rs [hunk 3]
```diff
@@ -138,3 +138,35 @@ proptest! {
...
+/// Test if sync lengths array with high values is not near tip.
+#[test]
+fn high_sync_lengths() {
+    let (status, mut recent_sync_lengths) = SyncStatus::new();
+
+    for _ in 0..RecentSyncLengths::MAX_RECENT_LENGTHS {
+        recent_sync_lengths.push_extend_tips_length(500);
+    }
+
+    assert!(status.is_close_to_tip() == false);
```
Comment (oxarbitrage) [inline zebrad/src/components/sync/status/tests.rs]:
```suggestion
    assert!(!status.is_close_to_tip());
```

clippy

**Review by oxarbitrage — COMMENTED** [review]
make clippy happy.

File: zebrad/src/components/sync/status/tests.rs [hunk 4]
```diff
@@ -138,3 +138,37 @@ proptest! {
...
+    assert!(status.is_close_to_tip());
+}
+
+/// Test if sync lengths array with high values is not near tip.
+#[test]
+fn high_sync_lengths() {
+    let (status, mut recent_sync_lengths) = SyncStatus::new();
+
+    /// The value 500 is based on the fact that sync lengths are around 500
+    /// blocks long when Zebra is syncing.
```
Comment (oxarbitrage) [inline zebrad/src/components/sync/status/tests.rs]:
```suggestion
    // The value 500 is based on the fact that sync lengths are around 500
    // blocks long when Zebra is syncing.
```

Another clippy, this ones should be normal code comments.

**Review by oxarbitrage — APPROVED** [review]
Tests look good to me, nice job.

**Pull request merged**